File: projects/11/JackCompiler/src/main/java/VMWriter.java
import java.io.*;

public class VMWriter {
    private final PrintWriter writer;

    // 构造函数：创建新的待写文件
    public VMWriter(String fileName) {
        // 打开文件
        try {
            writer = new PrintWriter(new BufferedWriter(new FileWriter(fileName)));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    // 写入VM push命令
    public void writePush(String segment, int index) {
        /*todo */
    }

    // 写入VM pop命令
    public void writePop(String segment, int index) {
        /*todo */
    }

    // 写入VM arithmetic命令
    public void writeArithmetic(String command) {
        /*todo */
    }

    // 写入VM label命令
    public void writeLabel(String label) {
        /*todo */
    }

    // 写入VM goto命令
    public void writeGoto(String label) {
        /*todo */
    }

    // 写入VM if-goto命令
    public void writeIf(String label) {
        /*todo */
    }

    // 写入VM call命令
    public void writeCall(String name, int nArgs) {
        /*todo */
    }

    // 写入VM function命令
    public void writeFunction(String name, int nVars) {
        /*todo */
    }

    // 写入VM return命令
    public void writeReturn() {
        /*todo */
    }

    // 关闭输出文件
    public void close() {
        writer.close();
    }
}

File: nand2TetrisCpp/JackCompilerCpp/src/VMWriter.cpp
#include "VMWriter.hpp"
#include <stdexcept>

VMWriter::VMWriter(const std::string &file_name) {
  out_.open(file_name);
  if (!out_)
    throw std::runtime_error("Cannot open output vm: " + file_name);
}
VMWriter::~VMWriter() { close(); }

void VMWriter::writePush(const std::string &segment, int index) {
  out_ << "push " << segment << " " << index << "\n";
}
void VMWriter::writePop(const std::string &segment, int index) {
  out_ << "pop " << segment << " " << index << "\n";
}
void VMWriter::writeArithmetic(const std::string &cmd) { out_ << cmd << "\n"; }
void VMWriter::writeLabel(const std::string &label) {
  out_ << "label " << label << "\n";
}
void VMWriter::writeGoto(const std::string &label) {
  out_ << "goto " << label << "\n";
}
void VMWriter::writeIf(const std::string &label) {
  out_ << "if-goto " << label << "\n";
}
void VMWriter::writeCall(const std::string &name, int nArgs) {
  out_ << "call " << name << " " << nArgs << "\n";
}
void VMWriter::writeFunction(const std::string &name, int nLocals) {
  out_ << "function " << name << " " << nLocals << "\n";
}
void VMWriter::writeReturn() { out_ << "return\n"; }
void VMWriter::close() {
  if (out_.is_open())
    out_.close();
}
